ipc/logger/pipeLogger.go

```diff
@@ -0,0 +1,171 @@
+package logger
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"io"
+	"os"
+)
+
+var _ Logger = (*PipeLogger)(nil)
+
+// LogMessage is a log message
+type LogMessage struct {
+	Level   LogLevel
+	Message string
+	Args    []interface{}
+}
+
+// PipeLogger is a pipe-based logger
+type PipeLogger struct {
+	pipe     *os.File
+	level    LogLevel
+	fallback Logger
+}
+
+// NewPipeLogger creates a new pipe logger
+func NewPipeLogger(level LogLevel, pipe *os.File) Logger {
+	return &PipeLogger{
+		level:    level,
+		pipe:     pipe,
+		fallback: NewDefaultLogger(level),
+	}
+}
+
+// Trace logs
+func (pipeLogger *PipeLogger) Trace(message string, args ...interface{}) {
+	if pipeLogger.level > LogTrace {
+		return
+	}
+
+	pipeLogger.sendMessage(&LogMessage{
+		Level:   LogTrace,
+		Message: message,
+		Args:    args,
+	})
+}
+
+// Debug logs
+func (pipeLogger *PipeLogger) Debug(message string, args ...interface{}) {
+	if pipeLogger.level > LogDebug {
+		return
+	}
+
+	pipeLogger.sendMessage(&LogMessage{
+		Level:   LogDebug,
+		Message: message,
+		Args:    args,
+	})
+}
+
+// Info logs
+func (pipeLogger *PipeLogger) Info(message string, args ...interface{}) {
+	if pipeLogger.level > LogInfo {
+		return
+	}
+
+	pipeLogger.sendMessage(&LogMessage{
+		Level:   LogInfo,
+		Message: message,
+		Args:    args,
+	})
+}
+
+// Warn logs
+func (pipeLogger *PipeLogger) Warn(message string, args ...interface{}) {
+	if pipeLogger.level > LogWarning {
+		return
+	}
+
+	pipeLogger.sendMessage(&LogMessage{
+		Level:   LogWarning,
+		Message: message,
+		Args:    args,
+	})
+
+}
+
+// Error logs
+func (pipeLogger *PipeLogger) Error(message string, args ...interface{}) {
+	if pipeLogger.level > LogError {
+		return
+	}
+
+	pipeLogger.sendMessage(&LogMessage{
+		Level:   LogError,
+		Message: message,
+		Args:    args,
+	})
+}
+
+func (pipeLogger *PipeLogger) sendMessage(message *LogMessage) {
+	payload, err := marshalJSON(message)
```

Comment:
make pipeLogger to have marshalizer.

Reply:
Fixed, but a bit different: serialization logic easy to change, well-encapsulated, but not as `marshalizer` interface & struct - but in a separate file instead. Otherwise, the caller of `NewPipeLogger` (Arwen's main module) must agree with the node part to use the same type of `marshalizer` also in the function `ReceiveLogThroughPipe` (which is not a member function) - that would result in unnecessary increase of complexity / over-engineering.

And we can also improve this when `pipeLogger` is moved to `core/logger` (outside `elrond-go`).

Reply:
at constructors arwen driver can give the marshalizer for any subprocess, subcomponent.

Reply:
Fixed. The only "not configurable" marshalizer remains the one used to transport initial arguments to Arwen through pipe. Since I removed command-line parameters altogether and only kept pipe parameters for simplicity, I didn't want to make other artifices to have this marshalizer configurable. But all others - the one for logs and the one for messages (the most important) are now configurable, received in constructors etc. :)